tracer/src/Datadog.Trace/DataStreamsMonitoring/DataStreamsManager.cs

```diff
@@ -0,0 +1,135 @@
+
+    public static DataStreamsManager Create(
+        ImmutableTracerSettings settings,
+        string defaultServiceName)
+        => new(settings.IsDataStreamsMonitoringEnabled, settings.Environment, defaultServiceName);
+
+    public Task DisposeAsync()
+    {
+        Volatile.Write(ref _isEnabled, false);
+        return Task.CompletedTask;
+    }
+
+    /// <summary>
+    /// Trys to extract a <see cref="PathwayContext"/>, from the provided <paramref name="headers"/>
+    /// If data streams is disabled, or no pathway is present, returns null.
+    /// </summary>
+    public PathwayContext? ExtractPathwayContext<TCarrier>(TCarrier headers)
+        where TCarrier : IBinaryHeadersCollection
+        => IsEnabled ? DataStreamsContextPropagator.Instance.Extract(headers) : null;
+
+    /// <summary>
+    /// Injects a <see cref="PathwayContext"/> into headers
+    /// </summary>
+    /// <param name="context">The pathway context to inject</param>
+    /// <param name="headers">The header collection to inject the headers into</param>
+    public void InjectPathwayContext<TCarrier>(PathwayContext? context, TCarrier headers)
+        where TCarrier : IBinaryHeadersCollection
+    {
+        if (!IsEnabled)
+        {
+            return;
+        }
+
+        if (context is not null)
+        {
+            DataStreamsContextPropagator.Instance.Inject(context.Value, headers);
+            return;
+        }
+
+        // This shouldn't happen normally, as you should call SetCheckpoint before calling InjectPathwayContext
+        // But if data streams was disabled, you call SetCheckpoint, and then data streams is enabled
+        // you will hit this code path
+        System.Diagnostics.Debug.Fail("Attempted to inject null pathway context");
+        Log.Debug("Attempted to inject null pathway context");
+    }
+
+    /// <summary>
+    /// Sets a checkpoint using the provided <see cref="PathwayContext"/>
+    /// NOTE: <paramref name="edgeTags"/> must be in correct sort order
+    /// </summary>
+    /// <param name="parentPathway">The current pathway</param>
+    /// <param name="edgeTags">Edge tags to set for the new pathway. MUST be sorted in alphabetical order</param>
+    /// <returns>If disabled, returns <c>null</c>. Otherwise returns a new <see cref="PathwayContext"/></returns>
+    public PathwayContext? SetCheckpoint(in PathwayContext? parentPathway, string[] edgeTags)
+    {
+        if (!IsEnabled)
+        {
+            return null;
+        }
+
+        try
+        {
+            var edgeStartNs = DateTimeOffset.UtcNow.ToUnixTimeNanoseconds();
+            var pathwayStartNs = parentPathway?.PathwayStart ?? edgeStartNs;
+
+            var nodeHash = HashHelper.CalculateNodeHash(_nodeHashBase, edgeTags);
+            var parentHash = parentPathway?.Hash ?? default;
+            var pathwayHash = HashHelper.CalculatePathwayHash(nodeHash, parentHash);
+
+            // TODO: send to aggregator
```

Comment:
Just commenting on TODO to make sure it wasn't forgotten (it might be handled in a later PR)

Reply:
Updated in https://github.com/DataDog/dd-trace-dotnet/pull/3191/files, thanks 🙂